### PR TITLE
mpi: autogen coll api

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Generating the collective APIs.

Slightly different from other API functions, the default "body of routines" will call `MPIR_Xxx` instead of `MPIR_Xxx_impl`. This is because collective uses `MPIR_Xxx` wrapper to check `CVAR`s and potentially do buffer swaps for gpu memory. Prototypes for `MPIR_Xxx_impl` is still automatically generated since `MPIR_Xxx` will always call `MPIR_Xxx_impl` in some branch.

The collective user buffer validations are by far the most complex part of the script logic. Because we have so many collectives, I believe having the logic dealt with script is overall a win in terms of maintenance.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/authorship

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Review by yfguo — APPROVED** [review]

**Comment by raffenet** [comment]
```
Slightly different from other API functions, the default "body of routines" will check cvars such as MPIR_CVAR_DEVICE_COLLECTIVES and per collective cvar such as MPIR_CVAR_BCAST_DEVICE_COLLECTIVE and either call MPID_Xxx or MPIR_Xxx_impl. This is the same as MPIR_Xxx functions. Having them directly generated made maintaining of MPIR_Xxx (one for each collectives) optional.
```
Is this true? I have checked a handful of the generated bindings, and they call `MPIR_xxx`. I see a note in the actual commit saying the `MPIR_xxx` routines are retained but not actually needed. But a `grep CVAR src/binding/c/coll/*` turns up empty.

**Comment by raffenet** [comment]
> > Slightly different from other API functions, the default "body of routines" will check cvars such as MPIR_CVAR_DEVICE_COLLECTIVES and per collective cvar such as MPIR_CVAR_BCAST_DEVICE_COLLECTIVE and either call MPID_Xxx or MPIR_Xxx_impl. This is the same as MPIR_Xxx functions. Having them directly generated made maintaining of MPIR_Xxx (one for each collectives) optional.
> 
> Is this true? I have checked a handful of the generated bindings, and they call `MPIR_xxx`. I see a note in the actual commit saying the `MPIR_xxx` routines are retained but not actually needed. But a `grep CVAR src/binding/c/coll/*` turns up empty.

Referring to this note in the commit message for `binding/c: autogen coll api`
```
Note that collectives always checks CVAR and decides whether to call
MPID_ funcitons or call MPIR_ impl functions. This logic is handled in
`MPIR_Xxx` (e.g. MPIR_Bcast). The script directly generates the same
logic without calling `MPIR_Xxx`.
```

**Comment by hzhou** [comment]
> 
> 
> > Slightly different from other API functions, the default "body of routines" will check cvars such as MPIR_CVAR_DEVICE_COLLECTIVES and per collective cvar such as MPIR_CVAR_BCAST_DEVICE_COLLECTIVE and either call MPID_Xxx or MPIR_Xxx_impl. This is the same as MPIR_Xxx functions. Having them directly generated made maintaining of MPIR_Xxx (one for each collectives) optional.
> 
> Is this true? I have checked a handful of the generated bindings, and they call `MPIR_xxx`. I see a note in the actual commit saying the `MPIR_xxx` routines are retained but not actually needed. But a `grep CVAR src/binding/c/coll/*` turns up empty.

The commit messages need to be updated. The original plan was to deprecate/remove MPIR_Xxx, but now we see that they are useful (e.g. gpu buffer swap), so I changed the plan. I'll update the messages, thanks for noticing.

File: src/binding/c/coll_api.txt [hunk 1]
```diff
@@ -0,0 +1,253 @@
+# vim: set ft=c:
+
+MPI_Allgather:
```
Comment (raffenet) [inline src/binding/c/coll_api.txt:3]: I don't see MPI_ERR_TYPE, MPI_ERR_BUFFER, MPI_ERR_COUNT, or MPI_ERR_OP for any of the generated collective validation.

**Review by raffenet — COMMENTED** [review]
Just the 2 standalone comments. Otherwise looks good!

File: src/binding/c/coll_api.txt (hunk 1 above)
Comment (hzhou) [reply src/binding/c/coll_api.txt:3]: Added commit "python: add missing error codes from user buffer validation" that fixes this.

**Comment by hzhou** [comment]
The commit message and PR description are updated.

test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
@raffenet Is this PR ready to merge?

**Review by raffenet — APPROVED** [review]

**Pull request merged**